hw02-my_array_list/src/main/java/ru/otus/collections/DIYArrayList.java

```diff
@@ -0,0 +1,288 @@
+package ru.otus.collections;
+
+import java.util.*;
+import java.util.function.Consumer;
+
+public class DIYArrayList<E> implements List<E> {
+    private int arraySize = 0;
+    private static final int DEFAULT_ARRAY_SIZE = 10;
+    transient Object[] array;
+    private static final Object[] EMPTY_ELEMENTDATA = {};
+
+    protected transient int modCount = 0;
+
+
+    public DIYArrayList(@org.jetbrains.annotations.NotNull Collection<? extends E> c) {
+        array = c.toArray();
+        if ((arraySize = array.length) != 0) {
+            if (array.getClass() != Object[].class)
+                array = Arrays.copyOf(array, arraySize, Object[].class);
+        } else {
+            this.array = EMPTY_ELEMENTDATA;
+        }
+    }
+
+    public DIYArrayList() {
+        this.arraySize = DEFAULT_ARRAY_SIZE;
+        this.array = new Object[DEFAULT_ARRAY_SIZE];
+    }
+
+    public DIYArrayList(int size) {
+        if (size > 0) {
+            this.arraySize = size;
+            this.array = new Object[size];
+        } else if (size == 0) {
+            this.array = EMPTY_ELEMENTDATA;
+        } else {
+            throw new IllegalArgumentException("Не правильное значение для размера массива");
+        }
+    }
+
+    @Override
+    public int size() {
+        return this.arraySize;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return (this.arraySize == 0);
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        throw new UnsupportedOperationException("Метод contains(Object o) не реализован");
+    }
+
+    @Override
+    public Iterator<E> iterator() {
+        return new Itr();
+    }
+
+    @Override
+    public Object[] toArray() {
+        return Arrays.copyOf(array, arraySize);
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        if (a.length < this.arraySize)
+            // Make a new array of a's runtime type, but my contents:
+            return (T[]) Arrays.copyOf(this.array, this.arraySize, a.getClass());
+        System.arraycopy(this.array, 0, a, 0, this.arraySize);
+        if (a.length > this.arraySize)
+            a[this.arraySize] = null;
+        return a;
+    }
+
+    @Override
+    public boolean add(E e) {
+        Boolean result = false;
```

Comment:
Расширяться каждый раз - очень дорого. Лучше скачкообразно по достижению заполнения. Посмотрите реализацию ArrayList - там это хорошо проработано. Это сделаем ArrayList действительно быстрым

Reply:
```suggestion
      private void grow(int minCapacity) {
        int oldArraySize = this.array.length;
        if (oldArraySize > 0 || this.array != EMPTY_ELEMENTDATA) {
            int newArraySize = newLength(oldArraySize,
                    minCapacity - oldArraySize,
                    oldArraySize >> 1);
            this.array = Arrays.copyOf(this.array,newArraySize);
        } else {
            this.array = new Object[Math.max(DEFAULT_ARRAY_SIZE, minCapacity)];
        }
    }

    @Override
    public boolean add(E e) {
        Boolean result = false;
        this.modCount++;
        if (this.arraySize == this.array.length)
            grow(this.arraySize + 1);
        this.array[this.arraySize] = e;
        this.arraySize++;
        result = true;
        return result;
    }    
```
Добавил скачкообразное увеличение

Reply:
👍🏻